File: Frontend/RetinaAI-Frontend/src/types/simulation.ts
/**
 * RetinaAI Phase 7 - Retinal Computational State-Space Simulation Types
 * 
 * Defines data structures for the LTI state-space simulation layer
 * downstream of Phase 6 biomarkers.
 * 
 * SCIENTIFIC INTEGRITY CONSTRAINT:
 * This is an engineering/research computational state model.
 * These metrics are computational states, NOT disease progression predictions,
 * clinical risk scores, or diagnostic severities.
 */

export interface BiomarkerInputSnapshot {
  vessel_density: number;
  branch_count: number;
  zone_b_count: number;
  distance_tortuosity: number;
  curvature_tortuosity: number;
  fractal_dimension: number;
  fractal_r_squared: number;
  [key: string]: any;
}

export interface NormalizedInputs {
  normalization_version: string;
  u_dens: number;
  u_branch: number;
  u_zb: number;
  u_tau_d: number;
  u_tau_c: number;
  u_Df_star: number;
  u_Df_raw: number;
  w_fit: number;
  [key: string]: any;
}

export interface SimulationOutputState {
  structural_complexity_state: number;
  tortuosity_computational_state: number;
  vascular_bed_density_state: number;
  composite_retinal_computational_state: number;
}

export interface SimulationTrajectory {
  time: number[];
  structural_complexity: number[];
  tortuosity: number[];
  vascular_bed_density: number[];
  composite_state: number[];
}

export type SimulationExecutionStatus = 
  | 'not_simulated'
  | 'pending'
  | 'running'
  | 'completed'
  | 'failed'
  | 'unavailable';

export interface RetinalSimulationRecord {
  id: string;
  screening_id: string;
  patient_id: string;
  eye: 'left' | 'right' | string;
  model_name: string;
  model_version: string;
  simulation_version: string;
  input_snapshot: BiomarkerInputSnapshot | null;
  normalized_inputs: NormalizedInputs | null;
  output_state: SimulationOutputState | null;
  trajectory: SimulationTrajectory | null;
  execution_status: SimulationExecutionStatus | string;
  runtime_seconds: number | null;
  error_message: string | null;
  created_at: string;
}

export interface SimulationSummaryCardProps {
  patientId: string;
  screenings?: any[];
  onNavigateToSimulation?: () => void;
}

File: Backend/retinaai-backend/schemas/simulation.py
"""
Pydantic schemas for the RetinaAI Simulink Computational Simulation API (Phase 7).

SCIENTIFIC DISCLAIMER:
Outputs represent engineering computational state metrics over dimensionless
relaxation time theta in [0, 10]. They are NOT clinical risk, DR grade, or
disease progression measures.
"""
from typing import Optional, Dict, Any, List
from datetime import datetime
from pydantic import BaseModel, Field, ConfigDict


class SimulateRequest(BaseModel):
    eye: str = Field("both", description="Eye to simulate: 'left', 'right', or 'both'")


class SimulationOutputState(BaseModel):
    structural_complexity_state: float = Field(..., ge=0.0, le=1.0)
    tortuosity_computational_state: float = Field(..., ge=0.0, le=1.0)
    vascular_bed_density_state: float = Field(..., ge=0.0, le=1.0)
    composite_retinal_computational_state: float = Field(..., ge=0.0, le=1.0)


class SimulationTrajectory(BaseModel):
    theta: List[float] = Field(default_factory=list)
    complexity_curve: List[float] = Field(default_factory=list)
    tortuosity_curve: List[float] = Field(default_factory=list)
    density_curve: List[float] = Field(default_factory=list)


class SimulationItem(BaseModel):
    id: str
    screening_id: str
    screening_display_id: Optional[str] = None
    patient_id: str
    patient_display_id: Optional[str] = None
    eye: str
    model_name: str = "retina_computational_state"
    model_version: str = "1.0.0"
    simulation_version: str = "1.0.0"
    contract_version: str = "1.0.0"
    execution_status: str  # 'completed', 'failed', 'disabled', 'invalid_input'
    input_snapshot: Dict[str, Any]
    normalized_inputs: Dict[str, Any]
    output_state: Optional[SimulationOutputState] = None
    trajectory: Optional[SimulationTrajectory] = None
    runtime_seconds: Optional[float] = None
    error_message: Optional[str] = None
    scientific_disclaimer: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class SimulationResponse(BaseModel):
    screening_id: str
    screening_display_id: Optional[str] = None
    patient_id: str
    patient_display_id: Optional[str] = None
    left: Optional[SimulationItem] = None
    right: Optional[SimulationItem] = None
    has_simulation: bool = False
    message: Optional[str] = None
